libr/arch/arch.c

```diff
@@ -0,0 +1,161 @@
+/* radare - LGPL - Copyright 2020 - pancake */
+
+#include <r_arch.h>
+#include <r_asm.h>
+#include "../config.h"
+
+static RArchPlugin *arch_static_plugins[] = { R_ARCH_STATIC_PLUGINS };
+
+static void plugin_free(RArchPlugin *p) {
+	if (p && p->fini) {
+		p->fini (NULL);
+	}
+}
+
+static bool is_valid(RArch *a, const char *name) {
+        RArchPlugin *h;
+        RListIter *iter;
+        if (!name || !*name) {
+                return false;
+        }
+        r_list_foreach (a->plugins, iter, h) {
+                if (!strcmp (h->name, name)) {
+                        return true;
+                }
+        }
+        return false;
+}
+
+R_API RArch *r_arch_new() {
+	RArch *a = R_NEW0 (RArch);
+	a->bits = R_SYS_BITS;
+	a->endian = R_SYS_ENDIAN_LITTLE;
+        a->plugins = r_list_newf ((RListFree)plugin_free);
+        if (!a->plugins) {
+                free (a);
+                return NULL;
+        }
+	size_t i;
+        for (i = 0; arch_static_plugins[i]; i++) {
+                r_arch_add (a, arch_static_plugins[i]);
+        }
+	return a;
+}
+
+R_API void r_arch_free(RArch *arch) {
+	free (arch);
+}
+
+// attributes
+
+static bool has_bits(RArchPlugin *h, RArchBits bits) {
+        return (h && h->bits && (bits & h->bits));
+}
+
+R_API bool r_arch_set_cpu(RArch *a, const char *cpu) {
+	// TODO: check if cpu is valid for the selected plugin
+	if (a) {
+		free (a->cpu);
+		a->cpu = cpu? strdup (cpu): NULL;
+	}
+	return true;
+}
+
+R_API bool r_arch_setup(RArch *a, const char *arch, RArchBits bits, RArchEndian endian) {
+	r_return_val_if_fail (a && arch, false);
+	bool ret = !r_arch_use (a, arch);
+	r_arch_set_endian (a, endian);
+	return ret | !r_arch_set_bits (a, bits);
+}
+
+R_API bool r_arch_set_syntax(RArch *a, int syntax) {
+	r_return_val_if_fail (a, false);
+	switch (syntax) {
+	case R_ASM_SYNTAX_REGNUM:
+	case R_ASM_SYNTAX_INTEL:
+	case R_ASM_SYNTAX_MASM:
+	case R_ASM_SYNTAX_ATT:
+	case R_ASM_SYNTAX_JZ:
+		a->syntax = syntax;
+		return true;
+	default:
+		return false;
+	}
+}
+
+R_API bool r_arch_set_endian(RArch *a, RArchEndian endian) {
```

Comment:
this should probably check whether the plugin supports such endian, otherwise it should fail.